backpack/custom_module/graph_utils.py

```diff
@@ -0,0 +1,203 @@
+                _change_node_to_module(
+                    node,
+                    "flatten",
+                    module,
+                    Flatten(start_dim, end_dim),
+                    (node.args[0],),
+                )
+                counter += 1
+
+    graph.lint()
+    print(f"\tFlatten transformed: {counter}")
+    return GraphModule(module, graph)
+
+
+def _transform_inplace_to_normal(
+    module: Module, initialize_recursion: bool = True
+) -> Module:
+    if initialize_recursion:
+        print("\tBegin transformation: in-place -> standard")
+        _transform_inplace_to_normal.counter = 0
+    if hasattr(module, "inplace") and module.inplace:
+        module.inplace = False
+        _transform_inplace_to_normal.counter += 1
+    for child_module in module.children():
+        _transform_inplace_to_normal(child_module, initialize_recursion=False)
+
+    if initialize_recursion:
+        print(f"\tIn-place changed: {_transform_inplace_to_normal.counter}")
+        delattr(_transform_inplace_to_normal, "counter")
+    return module
+
+
+def _transform_remove_duplicates(module: Module, max_depth: int = 100) -> GraphModule:
+    print("\tBegin transformation: remove duplicates")
+    counter = 0
+    graph: Graph = BackpackTracer().trace(module)
+    targets: Set[str] = set()
+    for node in graph.nodes:
+        if node.target in targets:
+            original_module = module.get_submodule(node.target)
+            for _ in original_module.parameters():
+                raise NotImplementedError(
+                    f"Transformation not successful, because a cycle was detected. "
+                    f"There is a module={original_module} with target={node.target} "
+                    f"that is used twice. This detected module has parameters."
+                )
+            new_module = deepcopy(original_module)
+            for i in range(max_depth):
+                target = f"{node.target}{i}"
+                try:
+                    module.get_submodule(target)
+                except AttributeError:
+                    module.add_submodule(target, new_module)
+                    node.target = target
+                    counter += 1
+                    break
+        else:
+            targets.add(node.target)
+
+    graph.lint()
+    print(f"\tDuplicates removed: {counter}")
+    return GraphModule(module, graph)
+
+
+def _change_node_to_module(
+    node: Node,
+    name: str,
+    base_module: Module,
+    new_module: Module,
+    args: tuple,
+) -> None:
+    max_depth = 100
+    for i in range(max_depth):
+        new_name = f"{name}_{i + 1}"
+        if hasattr(base_module, new_name):
+            continue
+        else:
+            break
+    name = new_name
+    if hasattr(base_module, name):
+        raise NotImplementedError(
+            f"There already exists a module named {name} in {base_module}. "
+            f"Consider increasing max_depth={max_depth} in the module naming space."
+        )
```

Comment:
Related to the magic number problem above.

Reply:
outdated